File: packages/python/kiff-guard/src/kiff_guard/guard.py
"""guard — the framework-agnostic core.

`Guard` knows nothing about any agent framework. It exposes three
primitives adapters build on, depending on the framework's control shape:

  observe(tool, args)            — learn + record an "observed" receipt.
                                   No decision, no run. (observe mode)
  decide_only(tool, args)        — learn + call KIFF decide and return the
                                   Decision. Does NOT run the tool and does
                                   NOT record — the adapter records exactly
                                   one receipt via record_executed (allowed)
                                   or record_withheld (withheld). (enforce)
  record_executed / record_withheld — the vote-shape adapter's single
                                   audit write, so one receipt per call.
  evaluate(tool, args, run)      — convenience for middleware frameworks
                                   that let the guard run the tool itself:
                                   observe-and-run, or decide-and-run-or-Hold
                                   (records one receipt internally).

Two adapter shapes use these differently:

  - Middleware / hook frameworks (Agno tool_hooks, Pydantic AI, …) let
    the guard run the tool, so they call `evaluate(tool, args, run=...)`.
  - Inverted-control / approval-native frameworks (Hermes pre_tool_call,
    LangGraph interrupt, OpenAI needs_approval, …) run the tool
    themselves after the hook returns; the hook only votes allow/block.
    They call `observe()` (observe mode) or `decide_only()` (enforce) and
    act on the returned Decision — never the run callback.

observe mode is decide-independent (#244): it works with no client and
no tenant, so a fresh user gets a real audit trail before they have any
KIFF account. The guard logic lives here, once; adapters add none.
"""

from __future__ import annotations

import time
from typing import Any, Callable, Dict, List, Optional

from .catalog import Catalog
from .client import Client
from .decision import Decision, Hold, Receipt


class Guard:
    """One Guard instance governs the tools of one logical agent. Share a
    Catalog + ledger across guards on the same tenant to get one tower,
    one learned surface, one audit log over every agent."""

    def __init__(
        self,
        client: Optional[Client] = None,
        tenant: str = "",
        agent: str = "agent",
        mode: str = "observe",
        catalog: Optional[Catalog] = None,
        ledger: Optional[List[Receipt]] = None,
    ):
        if mode not in ("observe", "enforce"):
            raise ValueError("mode must be 'observe' or 'enforce'")
        # enforce calls decide -> needs a client. observe is decide-
        # independent and works with no client at all (#244).
        if mode == "enforce" and client is None:
            raise ValueError("enforce mode requires a client")
        self.client = client
        self.tenant = tenant
        self.agent = agent
        self.mode = mode
        self.catalog = catalog if catalog is not None else Catalog()
        self.receipts: List[Receipt] = ledger if ledger is not None else []

    def observe(self, tool: str, args: Dict[str, Any]) -> None:
        """Record an observed receipt and learn the catalog. No decision,
        no run. The primitive for inverted-control adapters in observe
        mode (and the body of evaluate's observe branch).

        Decide-independent (#244): never calls KIFF. Valid with no client
        and no tenant."""
        self.catalog.record(self.agent, tool, args)
        self._record_observed(tool, args)

    def decide_only(self, tool: str, args: Dict[str, Any]) -> Decision:
        """Ask KIFF to decide and return the Decision WITHOUT running the
        tool and WITHOUT recording a receipt. The primitive for
        inverted-control adapters in enforce mode: the framework runs or
        skips the tool based on `decision.withheld`, then the adapter
        records exactly one receipt via `record_executed` (allowed) or
        `record_withheld` (withheld).

        Why no receipt here: the decision and the execution are two
        moments for a vote-shape adapter, but the *audit* must be one row
        per tool call — same shape as the middleware path. So recording
        is the adapter's explicit, single call, never a side effect of
        deciding. (Fixes the double-receipt issue: #239 / #250 review.)"""
        if self.client is None:
            raise ValueError("decide_only requires a client (enforce mode)")
        self.catalog.record(self.agent, tool, args)
        return self.client.decide(self.tenant, self.agent, tool, args)

    def evaluate(self, tool: str, args: Dict[str, Any], run: Callable[[], Any]) -> Any:
        """Convenience entry point for middleware frameworks that let the
        guard run the tool. `run` is a zero-arg callable that executes the
        tool (the adapter closes over the framework's continuation).
        Returns the tool result, or raises Hold in enforce mode when KIFF
        withholds clearance.

        Implemented on top of the observe / decide primitives so there is
        one source of truth for the observe/enforce + audit logic."""
        # Learn from every call, in both modes — integration is discovery.
        self.catalog.record(self.agent, tool, args)

        if self.mode == "observe":
            result = run()
            self._record_observed(tool, args)
            return result

        decision = self.client.decide(self.tenant, self.agent, tool, args)
        if decision.allowed:
            result = run()
            self._record_governed(tool, args, decision, executed=True)
            return result

        self._record_governed(tool, args, decision, executed=False)
        raise Hold(decision)

    def record_executed(self, tool: str, args: Dict[str, Any], decision: Decision) -> None:
        """Record exactly one governed receipt for an action the framework
        executed after an allowed `decide_only`. The vote-shape adapter's
        single audit write on the allowed path."""
        self._record_governed(tool, args, decision, executed=True)

    def record_withheld(self, tool: str, args: Dict[str, Any], decision: Decision) -> None:
        """Record exactly one governed receipt for an action KIFF withheld
        (the framework skipped it). The vote-shape adapter's single audit
        write on the withheld path. Pairs with `record_executed` so a
        vote-shape adapter emits one receipt per call, matching the
        middleware path."""
        self._record_governed(tool, args, decision, executed=False)

    def connect(
        self,
        adapter: str,
        project: str = "",
        environment: str = "",
        workflow: str = "",
        sdk_version: str = "",
    ) -> Any:
        """Opt into KIFF Cloud runtime discovery. Call this after creating
        the guard to register this runtime in the dashboard. Separate from
        observe/enforce so zero-config audit stays local unless the caller
        explicitly connects.

        Requires a client that implements GuardConnector (HTTPClient does).
        Returns the GuardConnection from the cloud."""
        from .client import GuardConnector

        if self.client is None:
            raise ValueError("connect requires a client")
        if not hasattr(self.client, "connect_guard"):
            raise ValueError("connect requires a client with connect_guard (HTTPClient)")
        return self.client.connect_guard(  # type: ignore[union-attr]
            agent_id=self.agent,
            adapter=adapter,
            mode=self.mode,
            project=project,
            environment=environment,
            workflow=workflow,
            sdk_version=sdk_version,
        )

    def save_draft(self, domain_name: str) -> Any:
        """Save the domain draft derived from observed traffic to the KIFF
        Cloud draft store, so it appears in the authoring UI. Renders the
        learned catalog with export_yaml and PUTs it.

        This is the credentialed half of instrument-first authoring; the
        credential-less fallback is to call export_yaml yourself and paste
        the result. Separate from observe/enforce so zero-config audit
        stays local unless the caller explicitly saves.

        Requires a client that implements DraftSaver (HTTPClient does).
        Returns the DraftResult from the cloud."""
        from .draft import export_yaml

        if self.client is None:
            raise ValueError("save_draft requires a client")
        if not hasattr(self.client, "save_draft"):
            raise ValueError("save_draft requires a client with save_draft (HTTPClient)")
        yaml_text = export_yaml(domain_name, self.catalog)
        return self.client.save_draft(yaml_text)  # type: ignore[union-attr]

    def observe_push(
        self,
        adapter: str,
        project: str = "",
        environment: str = "",
        workflow: str = "",
        sdk_version: str = "",
    ) -> Any:
        """Push the observed tool catalog to KIFF Cloud so it can derive a
        candidate domain from real traffic (POST /v1/guard/observations).
        Parallel to connect() / save_draft(): separate from observe/enforce
        so zero-config audit stays local unless the caller explicitly pushes.

        Turns the learned Catalog + the client's ToolMap bindings into tool
        observations. It only reports what is genuinely known:

          - name                — the observed tool.
          - parameter_schema    — the argument keys seen in real calls, as
                                  object properties (no invented types).
          - action/entity_type/entity_arg — from the ToolMap binding, when
                                  the tool is bound (real configuration).
          - observed_call_count — the real number of observed calls.

        It deliberately does NOT set `required`, a risk level, a state, or
        any threshold: those are human judgment, never inferred from
        traffic. Unbound/unmapped tools are still reported (name + observed
        args), carrying no action binding.

        Requires a client that implements ObservationPusher (HTTPClient does).
        Returns the GuardObservation echoed by the cloud."""
        from .client import GuardToolObservation

        if self.client is None:
            raise ValueError("observe_push requires a client")
        if not hasattr(self.client, "observe_guard"):
            raise ValueError("observe_push requires a client with observe_guard (HTTPClient)")

        tool_map = getattr(self.client, "tool_map", None)
        tools: List[GuardToolObservation] = []
        for name in sorted(self.catalog.tools):
            arg_keys = sorted(self.catalog.tools[name])
            binding = tool_map.get(name) if tool_map is not None else None
            count = self.catalog.counts.get(name, 0)
            tools.append(
                GuardToolObservation(
                    name=name,
                    parameter_schema=(
                        {"type": "object", "properties": {k: {} for k in arg_keys}}
                        if arg_keys
                        else None
                    ),
                    entity_arg=(binding.entity_arg if binding else None),
                    action=(binding.action if binding else None),
                    entity_type=(binding.entity_type if binding else None),
                    observed_call_count=(count if count else None),
                )
            )

        return self.client.observe_guard(  # type: ignore[union-attr]
            agent_id=self.agent,
            adapter=adapter,
            mode=self.mode,
            tools=tools,
            project=project,
            environment=environment,
            workflow=workflow,
            sdk_version=sdk_version,
        )

    # --- audit ---------------------------------------------------------

    def _record_observed(self, tool: str, args: Dict[str, Any]) -> None:
        self.receipts.append(
            Receipt(
                ts=time.time(), agent=self.agent, tool=tool, args=dict(args),
                outcome="observed", reason="observe mode: recorded, not governed",
                executed=True, state="observed",
            )
        )

    def _record_governed(self, tool: str, args: Dict[str, Any], decision: Decision, executed: bool) -> None:
        self.receipts.append(
            Receipt(
                ts=time.time(), agent=self.agent, tool=tool, args=dict(args),
                outcome=decision.outcome, reason=decision.reason,
                executed=executed, state="governed", proposal_id=decision.proposal_id,
            )
        )

File: packages/js/src/guard.ts
/**
 * guard — the framework-agnostic core.
 *
 * `Guard` knows nothing about any agent framework. It exposes the
 * primitives adapters build on, depending on the framework's control
 * shape (1:1 port of the Python SDK's guard.py):
 *
 *   observe(tool, args)           — learn + record an "observed" receipt.
 *                                   No decision, no run. (observe mode)
 *   decideOnly(tool, args)        — learn + call KIFF decide and return
 *                                   the Decision. Does NOT run the tool and
 *                                   does NOT record — the adapter records
 *                                   exactly one receipt via recordExecuted
 *                                   (allowed) or recordWithheld (withheld).
 *   recordExecuted / recordWithheld — the vote-shape adapter's single
 *                                   audit write, so one receipt per call.
 *   evaluate(tool, args, run)     — convenience for middleware frameworks
 *                                   that let the guard run the tool.
 *
 * OpenClaw's `before_tool_call` is a vote shape: OpenClaw runs the tool
 * itself; the hook returns a verdict. So its adapter uses observe() /
 * decideOnly() + recordExecuted/recordWithheld, never evaluate(run).
 *
 * observe mode is decide-independent (#244): it works with no client and
 * no tenant, so a fresh user gets a real audit trail before they have any
 * KIFF account. The guard logic lives here, once; adapters add none.
 */

import { Catalog } from "./catalog.js";
import type { Client, GuardConnection, GuardConnector } from "./client.js";
import { Decision, Hold, type Receipt } from "./decision.js";

export type GuardMode = "observe" | "enforce";

export interface GuardOptions {
  client?: Client;
  tenant?: string;
  agent?: string;
  mode?: GuardMode;
  /** share a Catalog across guards on one tenant for one learned surface. */
  catalog?: Catalog;
  /** share a ledger across guards for one audit log over every agent. */
  ledger?: Receipt[];
}

export interface GuardConnectOptions {
  /**
   * Adapter/runtime name, e.g. "openclaw". Required so Cloud can group
   * live agents by integration surface without inspecting tool traffic.
   */
  adapter: string;
  project?: string;
  environment?: string;
  workflow?: string;
  sdkVersion?: string;
}

export class Guard {
  readonly client?: Client;
  readonly tenant: string;
  readonly agent: string;
  readonly mode: GuardMode;
  readonly catalog: Catalog;
  readonly receipts: Receipt[];

  constructor(opts: GuardOptions = {}) {
    const mode = opts.mode ?? "observe";
    if (mode !== "observe" && mode !== "enforce") {
      throw new Error("mode must be 'observe' or 'enforce'");
    }
    // enforce calls decide -> needs a client. observe is decide-
    // independent and works with no client at all (#244).
    if (mode === "enforce" && !opts.client) {
      throw new Error("enforce mode requires a client");
    }
    this.client = opts.client;
    this.tenant = opts.tenant ?? "";
    this.agent = opts.agent ?? "agent";
    this.mode = mode;
    this.catalog = opts.catalog ?? new Catalog();
    this.receipts = opts.ledger ?? [];
  }

  /**
   * Record an observed receipt and learn the catalog. No decision, no
   * run. Decide-independent (#244): never calls KIFF. Valid with no
   * client and no tenant.
   */
  observe(tool: string, args: Record<string, unknown>): void {
    this.catalog.record(this.agent, tool, args);
    this.recordObserved(tool, args);
  }

  /**
   * Ask KIFF to decide and return the Decision WITHOUT running the tool
   * and WITHOUT recording a receipt. The primitive for vote-shape
   * adapters in enforce mode: the framework runs or skips the tool based
   * on `decision.withheld`, then the adapter records exactly one receipt
   * via recordExecuted (allowed) or recordWithheld (withheld).
   *
   * Why no receipt here: the decision and the execution are two moments
   * for a vote-shape adapter, but the *audit* must be one row per tool
   * call. So recording is the adapter's explicit, single call, never a
   * side effect of deciding. (The one-receipt rule, #239/#250.)
   */
  async decideOnly(tool: string, args: Record<string, unknown>): Promise<Decision> {
    if (!this.client) {
      throw new Error("decideOnly requires a client (enforce mode)");
    }
    this.catalog.record(this.agent, tool, args);
    return this.client.decide(this.tenant, this.agent, tool, args);
  }

  /**
   * Convenience entry point for middleware frameworks that let the guard
   * run the tool. `run` executes the tool. Returns the tool result, or
   * throws Hold in enforce mode when KIFF withholds clearance.
   */
  async evaluate<T>(tool: string, args: Record<string, unknown>, run: () => T | Promise<T>): Promise<T> {
    // Learn from every call, in both modes — integration is discovery.
    this.catalog.record(this.agent, tool, args);

    if (this.mode === "observe") {
      const result = await run();
      this.recordObserved(tool, args);
      return result;
    }

    const decision = await this.client!.decide(this.tenant, this.agent, tool, args);
    if (decision.allowed) {
      const result = await run();
      this.recordGoverned(tool, args, decision, true);
      return result;
    }
    this.recordGoverned(tool, args, decision, false);
    throw new Hold(decision);
  }

  /**
   * Opt into KIFF Cloud runtime discovery. This is separate from observe
   * and enforce so zero-config audit stays local unless the caller
   * provides a Cloud-capable client and calls connect().
   */
  async connect(opts: GuardConnectOptions): Promise<GuardConnection> {
    if (!isGuardConnector(this.client)) {
      throw new Error("connect requires a client with connectGuard");
    }
    return this.client.connectGuard({
      agentId: this.agent,
      adapter: opts.adapter,
      mode: this.mode,
      project: opts.project,
      environment: opts.environment,
      workflow: opts.workflow,
      sdkVersion: opts.sdkVersion,
    });
  }

  /**
   * Record exactly one governed receipt for an action the framework
   * executed after an allowed decideOnly. The vote-shape adapter's single
   * audit write on the allowed path.
   */
  recordExecuted(tool: string, args: Record<string, unknown>, decision: Decision): void {
    this.recordGoverned(tool, args, decision, true);
  }

  /**
   * Record exactly one governed receipt for an action KIFF withheld (the
   * framework skipped it). Pairs with recordExecuted so a vote-shape
   * adapter emits one receipt per call, matching the middleware path.
   */
  recordWithheld(tool: string, args: Record<string, unknown>, decision: Decision): void {
    this.recordGoverned(tool, args, decision, false);
  }

  // --- audit ---------------------------------------------------------

  private recordObserved(tool: string, args: Record<string, unknown>): void {
    this.receipts.push({
      ts: Date.now() / 1000,
      agent: this.agent,
      tool,
      args: { ...args },
      outcome: "observed",
      reason: "observe mode: recorded, not governed",
      executed: true,
      state: "observed",
    });
  }

  private recordGoverned(
    tool: string,
    args: Record<string, unknown>,
    decision: Decision,
    executed: boolean,
  ): void {
    this.receipts.push({
      ts: Date.now() / 1000,
      agent: this.agent,
      tool,
      args: { ...args },
      outcome: decision.outcome,
      reason: decision.reason,
      executed,
      state: "governed",
      proposalId: decision.proposalId,
    });
  }
}

function isGuardConnector(client: Client | undefined): client is Client & GuardConnector {
  return !!client && typeof (client as Partial<GuardConnector>).connectGuard === "function";
}
